### PR TITLE
feature(links, wallet-mobile): Deeplinks

### PR DESCRIPTION
**Pull request by stackchain**

### CODE REVIEW

File: .github/actions/wallet-mobile/action.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ inputs:
   node-version:
     description: "Node.js version to use"
     required: true
-    default: "16.19.0"
+    default: "18.19.1"
```
Comment (stackchain) [inline .github/actions/wallet-mobile/action.yml:8]: needed for eslint.

File: .tool-versions [hunk 2]
```diff
@@ -1,4 +1,4 @@
-nodejs 18.0.0
+nodejs 18.19.1
```
Comment (stackchain) [inline .tool-versions:1]: 18.18.0 is the minimum for eslint (not bumping the package cuz it will conflict with rn73 bump)

File: apps/wallet-mobile/src/AppNavigator.tsx [hunk 3]
```diff
@@ -136,7 +137,9 @@ export const AppNavigator = () => {
                 <Stack.Screen name="app-root">
                   {() => (
                     <SearchProvider>
-                      <WalletNavigator />
+                      <TransferProvider>
```
Comment (stackchain) [inline apps/wallet-mobile/src/AppNavigator.tsx:140]: lifted up the Transfer provider so we can populate the transfer at anytime (it is always reseted before navigating to in the normal flows) -> QR and Actions Banner.

File: apps/wallet-mobile/src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx [hunk 4]
```diff
@@ -48,14 +50,7 @@ export const WalletSelectionScreen = () => {
     onSuccess: ([wallet, walletMeta]) => {
       selectWalletMeta(walletMeta)
       selectWallet(wallet)
-
-      // fixes modal issue
```
Comment (stackchain) [inline apps/wallet-mobile/src/SelectedWallet/WalletSelection/WalletSelectionScreen.tsx:52]: this no longer applies, and should be done on callers.

File: apps/wallet-mobile/src/components/ConfirmTx/ConfirmTx.tsx [hunk 5]
```diff
@@ -269,6 +269,7 @@ export const ConfirmTx = ({
           {...buttonProps}
           disabled={isConfirmationDisabled || isProcessing || disabled}
           testID="confirmTxButton"
+          shelleyTheme
```
Comment (stackchain) [inline apps/wallet-mobile/src/components/ConfirmTx/ConfirmTx.tsx:272]: product requested long time ago

File: apps/wallet-mobile/src/features/Exchange/useCases/CreateExchangeOrderScreen/CreateExchangeOrderScreen.tsx [hunk 6]
```diff
@@ -38,9 +38,10 @@ export const CreateExchangeOrderScreen = () => {
   const {orderType, canExchange, providerId, provider, amount, referralLink: managerReferralLink} = useExchange()
 
   const providers = useExchangeProvidersByOrderType({orderType, providerListByOrderType: provider.list.byOrderType})
-  const providerSelected = Object.fromEntries(providers)[providerId]
-  const fee = providerSelected.supportedOrders?.[orderType]?.fee ?? 0
-  const Logo = providerSelected.id === 'banxa' ? BanxaLogo : EncryptusLogo
+  const providerSelected = new Map(providers).get(providerId)
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Exchange/useCases/CreateExchangeOrderScreen/CreateExchangeOrderScreen.tsx:41]: eslint was getting crazy, (current version is not compatible with typescript) only to suffice eslint check

File: apps/wallet-mobile/src/features/Links/common/constants.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import {freeze} from 'immer'
 
-export const knownApps: Readonly<Map<string, {name: string; vkey: string}>> = freeze(
+export const trustedApps: Readonly<Map<string, {name: string; vkey: string}>> = freeze(
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Links/common/constants.ts:3]: will come from the backend later signed by us

File: apps/wallet-mobile/src/features/Links/useCases/RequestedAdaPaymentWithLinkScreen/ShowDisclaimer/ShowDisclaimer.tsx
Comment (stackchain) [inline apps/wallet-mobile/src/features/Links/useCases/RequestedAdaPaymentWithLinkScreen/ShowDisclaimer/ShowDisclaimer.tsx:1]: needs product definition

File: apps/wallet-mobile/src/features/Scan/common/useTriggerScanAction.tsx [hunk 8]
```diff
@@ -65,16 +62,17 @@ export const useTriggerScanAction = ({insideFeature}: {insideFeature: ScanFeatur
...
       case 'send-only-receiver': {
-        navigateTo.back()
-        navigateTo.send()
-
-        if (insideFeature !== 'send') resetSendState()
+        if (insideFeature !== 'send') resetTransferState()
 
         receiverResolveChanged(scanAction.receiver)
+
+        navigateTo.startTransfer()
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Scan/common/useTriggerScanAction.tsx:75]: this fix the glitchy issue when going back on QR scan, navigation state was breaking the animation

Comment (stackchain) [reply apps/wallet-mobile/src/features/Scan/common/useTriggerScanAction.tsx:75]: @rahulnr7 this needs testing

File: apps/wallet-mobile/src/features/Send/useCases/ConfirmTx/Summary/ReceiverInfo.tsx [hunk 9]
```diff
@@ -26,7 +26,7 @@ export const ReceiverInfo = ({target}: Props) => {
         <>
           <View style={{flexDirection: 'row'}}>
             <Text style={styles.value}>
-              {receiver.selectedNameServer ? nameServerName[receiver.selectedNameServer] : ''}:
+              {receiver.selectedNameServer != null ? nameServerName[receiver.selectedNameServer] : ''}:
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Send/useCases/ConfirmTx/Summary/ReceiverInfo.tsx:29]: eslint was not complaining I'm a little bit scared, if there are other stuff like this around, this week we should bump rn and eslint/ts

File: packages/transfer/package.json [hunk 10]
```diff
@@ -139,13 +139,16 @@
     "@types/jest": "^29.5.12",
     "@types/react": "^18.2.55",
     "@types/react-test-renderer": "^18.0.7",
+    "@typescript-eslint/eslint-plugin": "^7.3.1",
```
Comment (stackchain) [inline packages/transfer/package.json]: ok, this was not supposed to be here cc: @stackchain 🤦🏻

File: yarn.lock [hunk 11]
```diff
@@ -2903,7 +2903,7 @@
     structured-headers "^0.4.1"
     tar "^6.0.5"
     tempy "^0.7.1"
-    terminal-link "^2.1.1"
+    terminal-link "2.1.1"
```
Comment (stackchain) [inline yarn.lock:2906]: hacky thing

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            if (redirectTo != null) redirectToChanged(decodeURIComponent(redirectTo))
+
+            const {address: receiver, amount, memo} = parsedCardanoLink.params
+            const ptAmount = Quantities.integer(asQuantity(amount ?? 0), decimals)
+            memoChanged(memo ?? '')
+            receiverResolveChanged(receiver ?? '')
+            amountChanged(ptAmount)
+            closeModal()
+            actionFinished()
+            navigateTo.startTransfer()
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx:42]: It would be nicer if the following was possible, but that's out of scope of the task

```
 navigateTo.startTransfer({memo, receiver, ptAmount})
```

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+      const handleOnContinue = () =>
+        startTransferWithLink(
+          {
+            info: {
+              version: 1,
+              feature: 'transfer',
+              useCase: 'request/ada-with-link',
+              params: params,
+            },
+            isTrusted: isTrusted,
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx]: `isTrusted,`

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+      Keyboard.dismiss()
+      const title = isTrusted ? strings.trustedPaymentRequestedTitle : strings.untrustedPaymentRequestedTitle
+      const handleOnContinue = () =>
+        startTransferWithLink(
+          {
+            info: {
+              version: 1,
+              feature: 'transfer',
+              useCase: 'request/ada-with-link',
+              params: params,
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx]: `params,`

File: apps/wallet-mobile/src/features/Links/common/useLinksShowActionResult.tsx [hunk 15]
```diff
@@ -0,0 +1,11 @@
+import {useLinks} from '@yoroi/links'
+import * as React from 'react'
+
+export const useLinksShowActionResult = () => {
+  const {action} = useLinks()
+  const initialRoute = action?.info.useCase === 'order/show-create-result' ? 'exchange-result' : 'wallet-selection'
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Links/common/useLinksShowActionResult.tsx:6]: I'd suggest `as const`

File: apps/wallet-mobile/src/WalletNavigator.tsx [hunk 16]
```diff
@@ -154,14 +155,14 @@ const WalletTabNavigator = () => {
 
 const Stack = createStackNavigator<WalletStackRoutes>()
 export const WalletNavigator = () => {
-  const initialRouteName = useInitialRouteName()
+  const {initialRoute} = useLinksShowActionResult()
+  useLinksRequestAction()
 
   // initialRoute doens't update the state of the navigator, only at first render
```
Comment (michalszorad) [inline apps/wallet-mobile/src/WalletNavigator.tsx]: typo

File: apps/wallet-mobile/src/features/Links/useCases/RequestedAdaPaymentWithLinkScreen/RequestedAdaPaymentWithLinkScreen.tsx [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        <ShowDisclaimer title={strings.disclaimer}>
+          <Text style={disclaimerStyle}>{description}</Text>
+        </ShowDisclaimer>
+
+        <Spacer height={16} />
+
+        {/* TODO: revisit SHOW the app name or unknown */}
+        {/* TODO: revisit SHOW verified / not verified icon and text */}
+        {/* TODO: revisit SHOW if it was initialized by Yoroi -> authorization */}
+        {/* TODO: revisit SHOW if it was initialized by Wallet -> walletId -> name */}
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Links/useCases/RequestedAdaPaymentWithLinkScreen/RequestedAdaPaymentWithLinkScreen.tsx:49]: Will it be revisited in another PR?

File: apps/wallet-mobile/src/features/Links/useCases/AskToOpenAWalletScreen/AskToOpenAWalletScreen.tsx [hunk 18]
```diff
@@ -0,0 +1,65 @@
+import {useLinks} from '@yoroi/links'
+import {useTheme} from '@yoroi/theme'
+import * as React from 'react'
+import {ScrollView, StyleSheet, Text, View, ViewProps} from 'react-native'
+import {SafeAreaView} from 'react-native-safe-area-context'
+
+import {Button, Spacer, useModal} from '../../../../components'
+import {useStrings} from '../../common/useStrings'
+
+export const AskToOpenWalletScreen = () => {
```
Comment (michalszorad) [inline apps/wallet-mobile/src/features/Links/useCases/AskToOpenAWalletScreen/AskToOpenAWalletScreen.tsx:10]: Is this just a placeholder for now, and will later allow users to choose a wallet?

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx (hunk 12 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx:42]: @michaeljscript not long time ago it was decided to move away from params, cuz they aren't safe, it always needs checking again on the Screens upon receiving them, that's why we favour context over params, yet, when exposing more routes to the word, we won't be able to avoid them anymore.

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx (hunk 13 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx]: make a suggestion, so I can merge it direcly

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx (hunk 14 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx]: ibid idem

File: apps/wallet-mobile/src/features/Links/common/useLinksShowActionResult.tsx (hunk 15 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Links/common/useLinksShowActionResult.tsx:6]: let me refactor is just a string, dunno why wrapped within a object 🤦🏻

File: apps/wallet-mobile/src/features/Links/useCases/RequestedAdaPaymentWithLinkScreen/RequestedAdaPaymentWithLinkScreen.tsx (hunk 17 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Links/useCases/RequestedAdaPaymentWithLinkScreen/RequestedAdaPaymentWithLinkScreen.tsx:49]: yap, we discussed after our sync today, a lot of work for product

File: apps/wallet-mobile/src/features/Links/useCases/AskToOpenAWalletScreen/AskToOpenAWalletScreen.tsx (hunk 18 above)
Comment (stackchain) [reply apps/wallet-mobile/src/features/Links/useCases/AskToOpenAWalletScreen/AskToOpenAWalletScreen.tsx:10]: it will work differently, cuz actions can start on mobile, (from a wallet), or can just start from a client (without any wallet), is also part of the product, if we receive a wallet later (that should come from mobile) we will open it for the user. For now if there is no wallet opened it will ask the user to open it.

File: apps/wallet-mobile/src/WalletNavigator.tsx (hunk 16 above)
Comment (stackchain) [reply apps/wallet-mobile/src/WalletNavigator.tsx]: make a suggestion

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx (hunk 13 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx]:
```suggestion
            isTrusted,
```

File: apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx (hunk 14 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/features/Links/common/useLinksRequestAction.tsx]:
```suggestion
              params,
```

File: apps/wallet-mobile/src/WalletNavigator.tsx (hunk 16 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/WalletNavigator.tsx]:
```suggestion
  // initialRoute doesn't update the state of the navigator, only at first render
```

**Review by michalszorad — APPROVED** [review]
